<center> <h1> Google Maps API </h> </center> 
<center> <h2> Homework 1 </h> </center> 
<center> <h3> Data Preprocessing </h> </center> 

A. Use the [*Nearby Seach*](https://developers.google.com/places/web-service/search) Method to find all the restaurant near to the UPY (20.988459, -89.736768) around 2km. In order to fulfill this item you need to:
* Create a *search_nearby_places* function with the appropiate parameters. It must return a JSON with the results. Places must be ranked by distance.
* Create a table with the following columns: Name, Place Id, Rating, Place Types, Total of User Ratings.

In [118]:
"""import the libraries to use"""
import pandas as pd 
import requests
import json

In [119]:
def search_nearby_places(key, location, radius, rankby=None):
    """In this function return a json with the restaurant that are to 2 km around of UPY 
    the arguments in this function is the key-> our API key, location->lat and lon of the UPY and rankby->None"""
    
    url = 'https://maps.googleapis.com/maps/api/place/nearbysearch/json?types=restaurant'
    
    parameters = {'key':key, 'location':location, 'radius':radius, 'rankby':rankby}
 
    
    res = requests.get(url, params=parameters)
    results = json.loads(res.content)
    return results

In [120]:
key = 'AIzaSyCpvCyAd9TA08ajzTW6QCU2m71H9MU55Qc'
location = '20.988459,-89.736768'
radius = 2000
rankby = None

In [121]:
places_json = search_nearby_places(key, location, radius, rankby)
places_json

{'html_attributions': [],
 'results': [{'geometry': {'location': {'lat': 20.9943362,
     'lng': -89.72616219999999},
    'viewport': {'northeast': {'lat': 20.99568418029151,
      'lng': -89.72480971970849},
     'southwest': {'lat': 20.9929862197085, 'lng': -89.72750768029151}}},
   'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/restaurant-71.png',
   'id': '51215063e260371f31be9c48b4b35f276677d8df',
   'name': 'La Glorieta (Cocina y Pizzería)',
   'opening_hours': {'open_now': True},
   'photos': [{'height': 4032,
     'html_attributions': ['<a href="https://maps.google.com/maps/contrib/112470977400009114281/photos">Óscar Albornoz</a>'],
     'photo_reference': 'CmRaAAAA5UIo0ykX1Eaon2v4cieFHDAkoPxc0dK0V7uYKxqZs2yEbps_rOwsjJB0-EgNf6fwrvzW6snzPSvLJeE1ghz1lR9FTdEPsuiiWAuM0wZrLt9m8pW2gps52B1gpypzCyKwEhDUgczeoagzb61AgqoFUs7yGhQ2to-ZwdRD-u1h0uGFn7Ki_3rHpA',
     'width': 3024}],
   'place_id': 'ChIJC9i7XEsNVo8Ro2-sdS45rBg',
   'plus_code': {'compound_code': 'X7VF+PG Mérida, Yu

In [122]:
df = pd.read_json(json.dumps(places_json['results']), orient = 'records')
df.head()

geometry  \
0  {'location': {'lat': 20.9943362, 'lng': -89.72...   
1  {'location': {'lat': 20.9892802, 'lng': -89.72...   
2  {'location': {'lat': 20.993318, 'lng': -89.723...   
3  {'location': {'lat': 20.9935837, 'lng': -89.72...   
4  {'location': {'lat': 20.9887511, 'lng': -89.72...   

                                                icon  \
0  https://maps.gstatic.com/mapfiles/place_api/ic...   
1  https://maps.gstatic.com/mapfiles/place_api/ic...   
2  https://maps.gstatic.com/mapfiles/place_api/ic...   
3  https://maps.gstatic.com/mapfiles/place_api/ic...   
4  https://maps.gstatic.com/mapfiles/place_api/ic...   

                                         id  \
0  51215063e260371f31be9c48b4b35f276677d8df   
1  0c95338760a6b57abb8c90f793757973d37a3858   
2  30608abf06a3d55762f3cf88c6ebfb59e6600755   
3  4084cb111246633c21ff68a31c522f01099d8f27   
4  752346a25e2f154673bf82a953e3efba988575a4   

                                          name        opening_hours  \
0              La Glorieta (Cocina y Pizzería)   {'open_now': True}   
1                                 Mr. Cazuelas                  NaN   
2                          El Rincón Del Tabas  {'open_now': False}   
3                               Moyoyos burger                  NaN   
4  Cocina economica "El sabor de doña antonia"  {'open_now': False}   

                                              photos  \
0  [{'height': 4032, 'html_attributions': ['<a hr...   
1                                                NaN   
2  [{'height': 3000, 'html_attributions': ['<a hr...   
3  [{'height': 1080, 'html_attributions': ['<a hr...   
4  [{'height': 363, 'html_attributions': ['<a hre...   

                      place_id  \
0  ChIJC9i7XEsNVo8Ro2-sdS45rBg   
1  ChIJ3WHHeJkMVo8RO_z2pfoDFZg   
2  ChIJn4UFHpwMVo8RZOfFHQCuIwo   
3  ChIJqxQ91h0NVo8R52sc8a1HlvE   
4  ChIJaQJm928NVo8RAN8mnDmvYEk   

                                           plus_code  rating  \
0  {'compound_code': 'X7VF+PG Mérida, Yucatan, Me...     4.1   
1  {'compound_code': 'X7QG+PH Mérida, Yucatan, Me...     4.0   
2  {'compound_code': 'X7VG+8C Mérida Municipality...     5.0   
3  {'compound_code': 'X7VG+CH Mérida, Yucatan, Me...     NaN   
4  {'compound_code': 'X7QH+G3 Mérida, Yucatan, Me...     NaN   

                     reference   scope  \
0  ChIJC9i7XEsNVo8Ro2-sdS45rBg  GOOGLE   
1  ChIJ3WHHeJkMVo8RO_z2pfoDFZg  GOOGLE   
2  ChIJn4UFHpwMVo8RZOfFHQCuIwo  GOOGLE   
3  ChIJqxQ91h0NVo8R52sc8a1HlvE  GOOGLE   
4  ChIJaQJm928NVo8RAN8mnDmvYEk  GOOGLE   

                                               types  user_ratings_total  \
0  [restaurant, food, point_of_interest, establis...                14.0   
1  [restaurant, food, point_of_interest, establis...                 3.0   
2  [restaurant, food, point_of_interest, establis...                 2.0   
3  [restaurant, food, point_of_interest, establis...                 NaN   
4  [restaurant, food, point_of_interest, establis...                 NaN   

                        vicinity  
0                         Mérida  
1               Calle 89, Caucel  
2                  Calle 77-A 77  
3                   Calle 77 800  
4  Calle 91 97314 Caucel, Mérida

In [123]:
relevant = ['name', 'place_id', 'rating', 'types', 'user_ratings_total'] # columns of our importance
df = df[relevant]

In [124]:
# use  dictionary for rename our columns
renombre = {'name': 'Name', 'place_id': 'Place id', 'rating':'Rating', 'types':'Types', 'user_ratings_total':'Total of User Ratings'}

In [125]:
df = df.rename(columns = renombre)
df

Name                     Place id  \
0               La Glorieta (Cocina y Pizzería)  ChIJC9i7XEsNVo8Ro2-sdS45rBg   
1                                  Mr. Cazuelas  ChIJ3WHHeJkMVo8RO_z2pfoDFZg   
2                           El Rincón Del Tabas  ChIJn4UFHpwMVo8RZOfFHQCuIwo   
3                                Moyoyos burger  ChIJqxQ91h0NVo8R52sc8a1HlvE   
4   Cocina economica "El sabor de doña antonia"  ChIJaQJm928NVo8RAN8mnDmvYEk   
5                   Cemitas Poblanas Doña Dulce  ChIJ-18misANVo8R8_gP-uVUAMY   
6                 Hamburguesas Maaaa de la 112.  ChIJXTh3gaoNVo8RovFdjQ-lN10   
7                                    Carnívoros  ChIJx9nxwp0MVo8RvujKFHXH6H4   
8                                    Modelorama  ChIJozhU3J0MVo8RKZTNG5qZyu4   
9                            La BOMBA DEL SABOR  ChIJXzbguNgLVo8RnxttdBhYakc   
10                                   La Carlota  ChIJVWRJiW8NVo8R7E46087W96E   
11                            The witch burgers  ChIJk8-rWwgNVo8RhVnpRaHabZ8   
12            Panuchos & Salbutes El Almendrito  ChIJ80ttQuMLVo8RC-oH1Eisi8A   
13                  HAMBURGUESAS HARAMYYU BIRIZ  ChIJHdPcMX0LVo8RhSB1GURFT7U   
14                                 PerryBurguer  ChIJldA29UgLVo8Rj6HAXz7I2iM   

    Rating                                              Types  \
0      4.1  [restaurant, food, point_of_interest, establis...   
1      4.0  [restaurant, food, point_of_interest, establis...   
2      5.0  [restaurant, food, point_of_interest, establis...   
3      NaN  [restaurant, food, point_of_interest, establis...   
4      NaN  [restaurant, food, point_of_interest, establis...   
5      4.8  [restaurant, food, point_of_interest, establis...   
6      3.7  [restaurant, food, point_of_interest, establis...   
7      4.1  [restaurant, food, point_of_interest, establis...   
8      4.8  [convenience_store, bar, liquor_store, restaur...   
9      5.0  [restaurant, food, point_of_interest, establis...   
10     5.0  [meal_delivery, restaurant, food, point_of_int...   
11     5.0  [restaurant, food, point_of_interest, establis...   
12     5.0  [restaurant, food, point_of_interest, establis...   
13     NaN  [restaurant, food, point_of_interest, establis...   
14     4.5  [restaurant, food, point_of_interest, establis...   

    Total of User Ratings  
0                    14.0  
1                     3.0  
2                     2.0  
3                     NaN  
4                     NaN  
5                    21.0  
6                     6.0  
7                    39.0  
8                     4.0  
9                     2.0  
10                    4.0  
11                    1.0  
12                    1.0  
13                    NaN  
14                   11.0

B. Create a function to Find the Place Id given Place Name (String) and Location (Lat/Lont Tuple).

In [126]:
def find_place(key, input_1, inputtype, locationbias=None):
    """This function return the ID of a place, if you want you can introduce lat and lon 
    the arguments: key->our API key, input_1->name our place, intputtype-> the type of input
    that we want in this case text, locationbias-> lat and long"""
    
    url = 'https://maps.googleapis.com/maps/api/place/findplacefromtext/json'
    if(locationbias!=None):
        parameters = {
            'key':key,
            'input':input_1,
            'inputtype':inputtype,
            'locationbias':str(locationbias)
        }
    else:
        parameters = {
            'key':key,
            'input':input_1,
            'inputtype':inputtype,
            'locationbias':None
        }
    res = requests.get(url, params = parameters)
    results = json.loads(res.content)
    return results['candidates'][0]['place_id']

inputtype = 'textquery'

C. Find the Place ID of the following places using your functions:
* McCarthy's Irish Pub - Caucel
* Starbucks Montejo
* Los Trompos Circuito

In [127]:
place_id_1 = find_place(key, "McCarthy's Irish Pub - Caucel", inputtype, 'POINT:20.9999339,-89.6847603')
place_id_1

'ChIJJyKKOKB0Vo8RCQcxkUw6hp8'

In [128]:
place_id_2 = find_place(key, "Starbucks Montejo", inputtype, 'POINT:20.984962, -89.6203225')
place_id_2

'ChIJaUUfylFxVo8RgEGjoansPbc'

In [129]:
place_id_3 = find_place(key, "Los Trompos Circuito", inputtype, 'POINT:20.9971994,-89.6240145')
place_id_3

'ChIJRwkizrxxVo8R1EULFLWZ2JQ'

D. Use the [*Place Details*](https://developers.google.com/places/web-service/details#PlaceDetailsResults) Method to find the Reviews of a Place in English and Spanish. In order to fulfill this item you need to:
* Create a *find_reviews* function with the appropiate parameters. It must return a JSON with the results. 
* Create a table with the following columns: Author Name, Language, Rating, Text, Time.

In [130]:
def find_reviews(key, place_id, language):
    """In this function return a json  that is util for find reviews
    arguments:key, place_id, language-> the language that we want"""
    
    url = "https://maps.googleapis.com/maps/api/place/details/json"
    paramaters={
        'key': key,
        'place_id':place_id,
        'language':language
    }
    
    res = requests.get(url, params = paramaters)
    result = json.loads(res.content)
    return result

language = 'es, en'

In [131]:
reviews_json = find_reviews(key, 'ChIJJyKKOKB0Vo8RCQcxkUw6hp8', language)
reviews_json

{'html_attributions': [],
 'result': {'address_components': [{'long_name': 'Caucel - Merida',
    'short_name': 'Caucel - Merida',
    'types': ['route']},
   {'long_name': 'Ciudad Caucel',
    'short_name': 'Cd Caucel',
    'types': ['sublocality_level_1', 'sublocality', 'political']},
   {'long_name': 'Mérida',
    'short_name': 'Mérida',
    'types': ['locality', 'political']},
   {'long_name': 'Yucatán',
    'short_name': 'Yuc.',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': 'Mexico',
    'short_name': 'MX',
    'types': ['country', 'political']},
   {'long_name': '97314', 'short_name': '97314', 'types': ['postal_code']}],
  'adr_address': 'Av. 23 Tablaje 40353, <span class="street-address">Caucel - Merida</span>, <span class="extended-address">Cd Caucel</span>, <span class="postal-code">97314</span> <span class="locality">Mérida</span>, <span class="region">Yuc.</span>, <span class="country-name">Mexico</span>',
  'formatted_address': 'Av. 23 Tablaje

In [132]:
reviews_json = pd.read_json(json.dumps(reviews_json['result']['reviews']), orient='records')

In [133]:
reviews_json = reviews_json.drop(columns=['author_url', 'profile_photo_url', 'relative_time_description'])

In [134]:
rename = {'author_name':'Author Name', 'language':'Language', 'rating':'Rating', 'text':'Text', 'time':'Time'}

In [135]:
reviews_json = reviews_json.rename(columns=rename)
reviews_json

Author Name Language  Rating  \
0  David Guillermo       en       3   
1          Patrick       en       5   
2            Islas       en       5   
3     Toni Aguilar       en       5   
4    monica ravell       en       5   

                                                Text        Time  
0  They need person to regulate the sound inside ...  1566612760  
1  One of the best pubs around.\nGreat service, g...  1518688093  
2  I love this place, one of the best places to d...  1522860496  
3  Rockers, music lovers in general will enjoy th...  1516852970  
4                 Great place to listen to good rock  1531876412

E. Find similar places to Los Trompos Circuito. Rank them by its "Prominence". In order to fulfill this item you need to:
* Create a *find_similar* function with the appropiate parameters. It must return a JSON with the results.
* Create a table with the following columns: Place Name, Plaece Id, Rating, Total of Use Ratings.

In [19]:
def find_similar(key, location, radius, rankby='prominence'):
    
    """This function return a json with information important for dataframe
    argumentos = key, loaction-> lat and long, radius->the radius that we want, rankby->prominence
    """
    url = 'https://maps.googleapis.com/maps/api/place/nearbysearch/json?types=restaurant'
    
    parameters = {
        'key':key,
        'location':location,
        'radius':radius,
        'rankby':rankby
    }
    res = requests.get(url, params=parameters)
    results = json.loads(res.content)
    return results 

key = 'AIzaSyCpvCyAd9TA08ajzTW6QCU2m71H9MU55Qc'
location = '20.9949269,-89.625473'
radius = 3000

In [20]:
a = find_similar(key, location, radius, rankby)
a

{'html_attributions': [],
 'next_page_token': 'CrQCIQEAAKgsW2nThZIvtXhQLWRS8Zdpz3qrgcob8RIRVhshhbu-9L7f4X1HlyJHZubfFEXHzjSOlzUFWMvk6k-rFPwzBMnOcpqu1gtyN0iR3N4JfdDeR7JppGrP1n6Ddaw_anypchAdbt-5OAxlOHkE2182AalF6fIu9233GxP8oR-T8w9BlvEtfsixMYjmMnzwHPkCRs_NPobXlCvRhDNeaohVQo6O5FlB1qHR64GqJJKluX11JBXmyo4TrL5oQ-8KtPIqI99h6XEH_aY6oXGnMWgb-iPOdcgSro5T9vZYvKxjPWwmiERv8XFEv4N-30PTPHLeEvIv-X2xa5KCIsPDDbR_gOIUIzKn46ASXXFJ2YUOWst5bNJtn729EMBFmwAzRbfl2pZp_9B9Nuv2xnz1x9s9DQgSEKEWnnOAHaziIMQzzQ41gjcaFGHaKnV_0NubFZpyM5Oavv3NuiCC',
 'results': [{'geometry': {'location': {'lat': 20.9716344, 'lng': -89.6255417},
    'viewport': {'northeast': {'lat': 20.9729234302915,
      'lng': -89.6242078197085},
     'southwest': {'lat': 20.97022546970849, 'lng': -89.62690578029151}}},
   'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/lodging-71.png',
   'id': '3ab345f3b5731f0dd4703e2df5e3d408149b0b35',
   'name': 'Hotel MedioMundo',
   'photos': [{'height': 1082,
     'html_attributions': ['<a href="https:/

In [21]:
a = pd.read_json(json.dumps(a['results']), orient='records')

In [22]:
relevant = ['name', 'place_id', 'rating', 'types', 'user_ratings_total'] # columns of our importance
a = a[relevant]
renombre = {'name': 'Name', 'place_id': 'Place id', 'rating':'Rating', 'types':'Types', 'user_ratings_total':'Total of User Ratings'}
a = a.rename(columns = renombre)
sorted_rating = a.sort_values(by='Rating', ascending=False)
df = sorted_rating.reset_index()
df = df.drop(columns=['index'])
df

Name                     Place id  \
0                        Hotel MedioMundo  ChIJ6UkVhN9zVo8R1l9m8TxbQPM   
1                              Sólo Pollo  ChIJRfu5cQN0Vo8RAIlYlD1h7nI   
2                                Eladio's  ChIJ9yoFwm5xVo8Rppx5yDFu_oQ   
3                                Muelle 8  ChIJY7vsm8Z2Vo8RJI7plSxAcmc   
4       El Vagón De La Baguette Campestre  ChIJi0sRC6V2Vo8R_gEG9tQArIM   
5                     Fausto's | Pizzeria  ChIJxeOAzV5xVo8RBoiqp5HsYt0   
6                     Pollo Brujo Reforma  ChIJw61ohuJzVo8RRd1vlJDdiBI   
7   Rosas & Xocolate Boutique Hotel + SPA  ChIJ68JYb1xxVo8RczcCV0_dsvs   
8                          Los 4 Hermanos  ChIJbW9TofBzVo8RuEZF-qA7sWA   
9                                  Subway  ChIJZ7rUWxpxVo8RmiC9ZdNmhmA   
10               Hotel Colonial de Mérida  ChIJvbQpe9cBzoURQu204GwtXw8   
11                  Benedetti's Pensiones  ChIJnY1i7o5zVo8RzJtdmRp-CcU   
12                     Domino's Campestre  ChIJqW-MvKZ2Vo8R74P9dOHTZ3Q   
13                  El Panucho de Kanasin  ChIJQ661ELB2Vo8R5Iuk2b2i0VQ   
14                            La Parrilla  ChIJLdBR77F2Vo8RuHJbN6nF_ng   
15                            Burger King  ChIJdeuSi_ZzVo8RhOx5yv6G6FU   
16                             McDonald's  ChIJV2Z-zFFxVo8RVVmNXyu_t_4   
17                   Hotel Montejo Palace  ChIJVT30Dp92Vo8R9yX_u-8nWBk   
18           El Guacamayo Restaurante Bar  ChIJS3UJ0FBxVo8RFjqPj7X_db0   
19                               Alcatraz  ChIJEWeemsl2Vo8R-1OGMd37lV0   

    Rating                                              Types  \
0      4.6  [lodging, restaurant, food, point_of_interest,...   
1      4.6  [meal_takeaway, restaurant, food, point_of_int...   
2      4.5  [bar, restaurant, food, point_of_interest, est...   
3      4.5  [restaurant, food, point_of_interest, establis...   
4      4.5  [restaurant, food, point_of_interest, establis...   
5      4.5  [restaurant, food, point_of_interest, establis...   
6      4.4  [restaurant, food, point_of_interest, establis...   
7      4.4  [restaurant, spa, lodging, food, gym, health, ...   
8      4.4  [restaurant, food, point_of_interest, establis...   
9      4.3  [restaurant, food, point_of_interest, establis...   
10     4.2  [lodging, restaurant, food, point_of_interest,...   
11     4.1  [meal_delivery, restaurant, food, point_of_int...   
12     4.1  [meal_delivery, meal_takeaway, restaurant, foo...   
13     4.1  [restaurant, food, point_of_interest, establis...   
14     4.1  [restaurant, food, point_of_interest, establis...   
15     4.1  [restaurant, food, point_of_interest, establis...   
16     4.0  [restaurant, food, point_of_interest, establis...   
17     3.8  [lodging, restaurant, food, point_of_interest,...   
18     3.6  [cafe, restaurant, food, point_of_interest, es...   
19     NaN  [restaurant, food, point_of_interest, establis...   

    Total of User Ratings  
0                    81.0  
1                    12.0  
2                  2061.0  
3                   197.0  
4                   119.0  
5                   661.0  
6                   229.0  
7                   518.0  
8                   260.0  
9                   580.0  
10                  587.0  
11                  157.0  
12                  359.0  
13                  165.0  
14                 1181.0  
15                 2082.0  
16                  906.0  
17                   40.0  
18                    5.0  
19                    NaN

F. Complete all the above tasks but using the Python [GoogleMaps](https://github.com/googlemaps/google-maps-services-python) Library. [Here](https://googlemaps.github.io/google-maps-services-python/docs/#module-googlemaps) you can find the documentation.

## A

In [55]:
import googlemaps
gmaps = googlemaps.Client(key = 'AIzaSyCpvCyAd9TA08ajzTW6QCU2m71H9MU55Qc')

In [56]:
location = '20.988459, -89.736768'
radius = 2000
type_1 = 'restaurant'

In [57]:
nearby_places = gmaps.places_nearby(location, radius, type_1, open_now = False)
dd1 = pd.read_json(json.dumps(nearby_places['results']), orient = 'columns')
dd1[['name','place_id','rating','types','user_ratings_total']]

name  \
0                  Restaurant El Parguito De Celestún   
1                                 Asadero Doña Mary 2   
2                                    La Pasta Rodante   
3                                     Pizzeria D'Alex   
4                   Panuchos & Salbutes El Almendrito   
5                               Taquería "Miguelito".   
6                       Hamburguesas Maaaa de la 112.   
7                                       Lady's Hochos   
8                                      Casa De Oliver   
9                     La Glorieta (Cocina y Pizzería)   
10                                 Antojitos El Cielo   
11                                      PANUCHOS DANI   
12                                       Pizzeria Max   
13                                         El Cerrito   
14                          Taquería "Méndez Chiapas"   
15                                        Burguer Fer   
16                            Tropitacos "El Caporal"   
17                                 "TAMALES DEL CHEF"   
18  ANTOJITOS MEXICANOS AMIRANY (sopes, quesadilla...   
19                                Asadero Doña Mary 2   

                       place_id  rating  \
0   ChIJi6qRzU4NVo8RCHrvRJcvtyU     0.0   
1   ChIJZVDHeQgNVo8Rx2ccWv7mAOI     0.0   
2   ChIJO5132EILVo8R7kJZAL-_w-0     5.0   
3   ChIJq8dvBqAMVo8RRRwkEePmmss     4.9   
4   ChIJ80ttQuMLVo8RC-oH1Eisi8A     5.0   
5   ChIJ8dMBypANVo8RiedlO3x2iTM     4.6   
6   ChIJXTh3gaoNVo8RovFdjQ-lN10     3.7   
7   ChIJaQYRU6YLVo8RBSYWcVxecZE     5.0   
8   ChIJW6FbE6AMVo8R-pDCrwCDRgU     3.9   
9   ChIJC9i7XEsNVo8Ro2-sdS45rBg     4.1   
10  ChIJKSZhQVwLVo8RiNqp6upowD4     4.5   
11  ChIJe8W_jl4LVo8RRnJ2wqJEicY     4.4   
12  ChIJlXjObdgLVo8ReOxQTCvDNjw     4.6   
13  ChIJ3WIPkKAMVo8Ra0lEinl-tjw     4.6   
14  ChIJo23VAsMLVo8RaisTtrmFeu4     4.4   
15  ChIJJROnL1wLVo8R8Flmf86aOl4     3.0   
16  ChIJYfaQLEsLVo8RfE4wa22iEWk     5.0   
17  ChIJ0_HSEVwLVo8RcvWe4fG4t4o     0.0   
18  ChIJQSdKyWcNVo8RhrqoqOeSXtg     5.0   
19  ChIJ76OOP6AMVo8RDqYeLvM15dA     5.0   

                                                types  user_ratings_total  
0   [restaurant, food, point_of_interest, establis...                   0  
1   [restaurant, food, point_of_interest, establis...                   0  
2   [restaurant, food, point_of_interest, establis...                   7  
3   [restaurant, food, point_of_interest, establis...                   9  
4   [restaurant, food, point_of_interest, establis...                   1  
5   [restaurant, food, point_of_interest, establis...                  76  
6   [restaurant, food, point_of_interest, establis...                   6  
7   [restaurant, food, point_of_interest, establis...                   6  
8   [restaurant, food, point_of_interest, establis...                  23  
9   [restaurant, food, point_of_interest, establis...                  14  
10  [restaurant, food, point_of_interest, establis...                  26  
11  [restaurant, food, point_of_interest, establis...                 379  
12  [restaurant, food, point_of_interest, establis...                  14  
13  [restaurant, food, point_of_interest, establis...                  19  
14  [restaurant, food, point_of_interest, establis...                 139  
15  [restaurant, food, point_of_interest, establis...                   2  
16  [restaurant, food, point_of_interest, establis...                   3  
17  [restaurant, food, point_of_interest, establis...                   0  
18  [restaurant, food, point_of_interest, establis...                   2  
19  [restaurant, food, point_of_interest, establis...                   1

## B

In [92]:
def find_with_g(loc, restaurant, radius=2000, intputtype='textquery'):
    g = gmaps.find_place(input=restaurant,input_type=inputtype, location_bias=loc)
    results =g ['candidates'][0]['place_id']
    
    return results

## C

In [93]:
McCarthys = find_with_g('point:20.9950689,-89.6975387',"McCarthy\'s Irish Pub - Caucel")
print(f'McCarthys id = {McCarthys}')

McCarthys id = ChIJJyKKOKB0Vo8RCQcxkUw6hp8


In [95]:
Starbucks = find_with_g('point:20.984962,-89.6203225', "Starbucks Montejo")
print(f'Starbucks id = {Starbucks}')

Starbucks id =ChIJaUUfylFxVo8RgEGjoansPbc


In [97]:
los_trompos = find_with_g('point:20.9971994,-89.6240145', 'Los Trompos Circuito')
print(f'Los trompos circuito id = {los_trompos}')

Los trompos circuito id = ChIJpVKXOq12Vo8RK4bUWILT9dI


## D

In [115]:
def details_json(place_id, lang):
    g = gmaps.place(place_id=place_id, language=lang)
    results = g['result']['reviews']
    
    return results

In [116]:
b = details_json('ChIJpVKXOq12Vo8RK4bUWILT9dI', 'es,en')
b

[{'author_name': 'Patricia Ramirez',
  'author_url': 'https://www.google.com/maps/contrib/102399857172028830731/reviews',
  'language': 'en',
  'profile_photo_url': 'https://lh5.ggpht.com/-j8h9-WtknfA/AAAAAAAAAAI/AAAAAAAAAAA/sK8ulwqNrCU/s128-c0x00000000-cc-rp-mo/photo.jpg',
  'rating': 5,
  'relative_time_description': '3 weeks ago',
  'text': "Best tacos I've ever had!",
  'time': 1567825418},
 {'author_name': 'andres dueck',
  'author_url': 'https://www.google.com/maps/contrib/110024682673670197390/reviews',
  'language': 'en',
  'profile_photo_url': 'https://lh3.ggpht.com/-kfgVnS2Rgo8/AAAAAAAAAAI/AAAAAAAAAAA/5qpBKQ1XUR4/s128-c0x00000000-cc-rp-mo/photo.jpg',
  'rating': 5,
  'relative_time_description': 'in the last week',
  'text': 'Food was extremely good!! Worse thing was I overate',
  'time': 1569549855},
 {'author_name': 'Robert Hamblin',
  'author_url': 'https://www.google.com/maps/contrib/113255653587053238945/reviews',
  'language': 'en',
  'profile_photo_url': 'https://lh5.g

In [117]:
b = pd.read_json(json.dumps(b), orient='records')
b[['author_name', 'language', 'rating', 'text', 'time']]

author_name language  rating  \
0  Patricia Ramirez       en       5   
1      andres dueck       en       5   
2    Robert Hamblin       en       5   
3       Cris Martin       en       5   
4     Christi Foist       en       5   

                                                text        time  
0                          Best tacos I've ever had!  1567825418  
1  Food was extremely good!! Worse thing was I ov...  1569549855  
2  Friendly staff. Clean bathrooms. Have the Alam...  1561151931  
3       Great mexican food frijoles charos are great  1560024294  
4        Great, affordable tacos and friendly staff.  1563161376

## E

In [136]:
def similar_g(loca, radius, type_2='restaurant', rankby='prominence'):
    results = gmaps.places_nearby(location=loca, radius=radius, type=type_2, rank_by=rankby)
    
    return results

In [139]:
similar_g('20.9949269,-89.625473', 2000)

{'html_attributions': [],
 'next_page_token': 'CrQCIwEAAFfwcXYgp3kmYzl8Ilhw0xzDXxaEB0PteAaJiyfr5bcbZ42taTSILL9YnZnPjoabLpF-RY3iExU46LxTBjs_CqSsBb0sqa64QNBBHbKv43uGpK0hc-4snULEHXeRT4Ri3-Mc6jzdFWxHedrDxMGCemu1eT4ka9spWl0Zo_E6GIhit16OTB_VdsMq00kyvjKIKAqk9Us-yqK_e-FLUQK7PbqKSORIV1-oxvb3cRF4mJ1jusH0bThcjW-QaENsrJSrQ_6lQfVWQETTSmJKEGm3J0VdI36h_XHwczbaC0x4rcAUPmBI2R1G5ffjCWUQK0phcx-AESD5l_KbH9EZMdpHmaDm3Oe1P0rVSHtBTaH3xDOw32qxRQMp5cWluFubZQPn8ajZSFVaYT1zCJC19ma89wcSEH_64cayJDvJrjKr5HmzAhcaFL8oztFIwb39MouJc1Ns7qRoJbz1',
 'results': [{'geometry': {'location': {'lat': 20.979003,
     'lng': -89.61842399999999},
    'viewport': {'northeast': {'lat': 20.9803818802915,
      'lng': -89.6172604697085},
     'southwest': {'lat': 20.9776839197085, 'lng': -89.6199584302915}}},
   'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/restaurant-71.png',
   'id': '1f33d6c7ef7b8b1c060d3bfd055c2fa887f39879',
   'name': 'Rosas & Xocolate Boutique Hotel + SPA',
   'opening_hours': {'open_now': False},
 

In [141]:
similar_d = similar_g('20.9949269,-89.625473', 2000)

dd3 = pd.read_json(json.dumps(similar_d['results']),orient='records')
dd3 = dd3[['name', 'place_id', 'rating','user_ratings_total']]
dd3 = dd3.sort_values(by='rating', ascending=False)
dd3.reset_index(drop=True)

name                     place_id  \
0                     Taqueria Villaldana  ChIJI23PVwZ0Vo8RiU0kQI6SebU   
1                              Sólo Pollo  ChIJRfu5cQN0Vo8RAIlYlD1h7nI   
2             Los Originales Tacos de Res  ChIJScV8m1JxVo8RGtH_DNMAHcs   
3                             Los Trompos  ChIJpVKXOq12Vo8RK4bUWILT9dI   
4                        Cocina de Mirtya  ChIJcwOYpvtzVo8Rd98mB5wi61o   
5                                Eladio's  ChIJ9yoFwm5xVo8Rppx5yDFu_oQ   
6                            La Tradicion  ChIJ3cHOYVlxVo8RGvQdcRllqFE   
7                   Pizza del Perro Negro  ChIJS67Ey6x2Vo8RTiTblDl7VT0   
8   Rosas & Xocolate Boutique Hotel + SPA  ChIJ68JYb1xxVo8RczcCV0_dsvs   
9                          Los 4 Hermanos  ChIJbW9TofBzVo8RuEZF-qA7sWA   
10                    Pollo Brujo Reforma  ChIJw61ohuJzVo8RRd1vlJDdiBI   
11                            Los 7 Mares  ChIJ6zLpmP9zVo8R0Al6847lN_s   
12                       The Sausage Lady  ChIJm0u1w1dxVo8RNq6sVo0Mzd4   
13                        Mi Viejo Molino  ChIJA2XvLq12Vo8RmHzVkbC-a1M   
14                              Pizza Hut  ChIJJX34FnFzVo8RekHn1Byql38   
15                            La Parrilla  ChIJLdBR77F2Vo8RuHJbN6nF_ng   
16                             McDonald's  ChIJV2Z-zFFxVo8RVVmNXyu_t_4   
17                   Hotel Montejo Palace  ChIJVT30Dp92Vo8R9yX_u-8nWBk   
18           El Guacamayo Restaurante Bar  ChIJS3UJ0FBxVo8RFjqPj7X_db0   
19           Alvéolo - Panadería & Cocina  ChIJhZ1K5wB0Vo8R2ErYyB9SGzY   

    rating  user_ratings_total  
0      4.7                54.0  
1      4.6                12.0  
2      4.5                33.0  
3      4.5              3116.0  
4      4.5                 2.0  
5      4.5              2061.0  
6      4.4              1059.0  
7      4.4              1787.0  
8      4.4               518.0  
9      4.4               260.0  
10     4.4               230.0  
11     4.4                87.0  
12     4.3                10.0  
13     4.2              1383.0  
14     4.1               500.0  
15     4.1              1181.0  
16     4.0               906.0  
17     3.8                40.0  
18     3.6                 5.0  
19     NaN                 NaN

G. Create a Github Project as a Tutorial about How to Use the Google API. Don't forget to include a section with the set of instructions to obtain the Google API Key. In this tutorial, compare the use of the Google Maps [Places API] (https://developers.google.com/places/web-service/intro) and the Python [GoogleMaps](https://github.com/googlemaps/google-maps-services-python) Library using this activity tasks as examples.

## Checklist

| Task | Points |
|:----:|:------:|
|   Class Participation  |   10   |
|   a  |   10   |
|   b  |   10   |
|   c  |   10   |
|   d  |   10   |
|   e  |   10   |
|   f  |   20   |
|   g  |   20   |


* **Important!** Partial credit can be obteined in each task.
* Don not forget to document your functions! [Here](http://google.github.io/styleguide/pyguide.html) is the Python Programming Styleguide from Google. [Section 3.8](http://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings) is about Comments and Docstrings.
* To obtain the Class Participation Score you must complete an early submission with tasks: A, B and C.